labelindex/named_port_index.go

```diff
@@ -306,7 +308,29 @@ func extractCIDRsFromNetworkSet(netSet *model.NetworkSet) []ip.CIDR {
 	a := netSet.Nets
 	combined := make([]ip.CIDR, 0, len(a))
 	for _, addr := range a {
-		combined = append(combined, ip.CIDRFromCalicoNet(addr))
+		cidr := ip.CIDRFromCalicoNet(addr)
+		if cidr.Prefix() == 0 {
+			// Special case: the linux dataplane can't handle 0-length CIDRs so we split it into
+			// multiple CIDRs.  Note: if the /1s were also in the network set, the deduplication is
+			// handled by reference counting in the ipSetData struct.
+			log.Debug("Converting 0 length CIDR to pair of /1s")
+			switch cidr.Version() {
+			case 4:
+				combined = append(combined,
+					ip.MustParseCIDROrIP("0.0.0.0/1"),
+					ip.MustParseCIDROrIP("128.0.0.0/1"),
+				)
+			case 6:
+				combined = append(combined,
+					ip.MustParseCIDROrIP("::/1"),
+					ip.MustParseCIDROrIP("8000::/1"),
+				)
+			default:
+				log.WithField("cidr", cidr).Panic("Unknown IP version")
+			}
+			continue
+		}
+		combined = append(combined, cidr)
```

Comment:
For me this would be clearer in an `else` branch.  (I know it's OK as is because of the `continue`, but it took me a while to see that.)